src/evmc-client/src/agent.rs

```diff
@@ -0,0 +1,87 @@
+mod identity;
+
+use std::path::{Path, PathBuf};
+
+use candid::utils::ArgumentEncoder;
+use candid::{encode_args, CandidType, Principal};
+use did::codec;
+use ic_agent::identity::PemError;
+use ic_exports::ic_kit::RejectionCode;
+use serde::Deserialize;
+use thiserror::Error;
+
+use crate::client::EvmCanisterClient;
+use crate::IcResult;
+
+#[derive(Error, Debug)]
+pub enum AgentError {
+    #[error("agent error: {0}")]
+    Agent(#[from] ic_agent::AgentError),
+
+    #[error("failed to read PEM file {0}: {1}")]
+    PemError(PathBuf, PemError),
+}
+
+pub type Result<T> = std::result::Result<T, AgentError>;
+
+pub struct IcAgentClient {
+    canister_id: Principal,
+    agent: ic_agent::Agent,
+}
+
+impl IcAgentClient {
+    /// Initialize an IC Agent with a PEM file
+    pub async fn with_identity(
+        canister: Principal,
+        identity_path: &Path,
+        network: &str,
+    ) -> Result<Self> {
+        let agent = identity::init_agent(identity_path, network).await?;
+        Ok(Self {
+            canister_id: canister,
+            agent,
+        })
+    }
+
+    /// Initialize an IC Agent with an existing agent
+    pub fn with_agent(canister: Principal, agent: ic_agent::Agent) -> Self {
+        Self {
+            canister_id: canister,
+            agent,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl EvmCanisterClient for IcAgentClient {
+    async fn query<T, R>(&self, method: &str, args: T) -> IcResult<R>
+    where
+        T: ArgumentEncoder + Send,
+        R: for<'de> Deserialize<'de> + CandidType,
+    {
+        let args = encode_args(args).expect("encode args failed");
```

Comment:
Same as for the other client, I guess it is better not to panic

Reply:
Done